synapse/storage/data_stores/main/events_worker.py

```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    @cached(tree=True, cache_context=True)
+    async def get_unread_message_count_for_user(
+        self, room_id: str, user_id: str, cache_context: _CacheContext,
+    ):
+        last_read_event_id = await self.get_last_receipt_event_id_for_user(
+            user_id=user_id,
+            room_id=room_id,
+            receipt_type="m.read",
+            on_invalidate=cache_context.invalidate,
+        )
+
+        return await self.db.runInteraction(
+            "get_unread_message_count_for_user",
+            self._get_unread_message_count_for_user_txn,
+            user_id,
+            room_id,
+            last_read_event_id,
+        )
+
+    def _get_unread_message_count_for_user_txn(
+        self,
+        txn: LoggingTransaction,
```

Comment:
Cursor is probably preferable here, unless you're particularly relying on the methods in LoggingTransaction